plugins/auth-backend-module-oidc-provider/src/authenticator.ts

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright 2023 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  Issuer,
+  ClientAuthMethod,
+  TokenSet,
+  UserinfoResponse,
+  Strategy as OidcStrategy,
+} from 'openid-client';
+import {
+  createOAuthAuthenticator,
+  PassportOAuthAuthenticatorHelper,
+  PassportOAuthDoneCallback,
+} from '@backstage/plugin-auth-node';
+
+/** @public */
+export const oidcAuthenticator = createOAuthAuthenticator({
+  defaultProfileTransform:
+    PassportOAuthAuthenticatorHelper.defaultProfileTransform,
+  async initialize({ callbackUrl, config }) {
+    const clientId = config.getString('clientId');
+    const clientSecret = config.getString('clientSecret');
+    const metadataUrl = config.getString('metadataUrl');
+    const customCallbackUrl = config.getOptionalString('callbackUrl');
+    const callbackUrl2 = customCallbackUrl || callbackUrl;
+    const tokenEndpointAuthMethod = config.getOptionalString(
+      'tokenEndpointAuthMethod',
+    ) as ClientAuthMethod;
+    const tokenSignedResponseAlg = config.getOptionalString(
+      'tokenSignedResponseAlg',
+    );
+    const initializedScope = config.getOptionalString('scope');
+    const initializedPrompt = config.getOptionalString('prompt');
+
+    const issuer = await Issuer.discover(metadataUrl);
+    const client = new issuer.Client({
+      access_type: 'offline', // this option must be passed to provider to receive a refresh token
+      client_id: clientId,
+      client_secret: clientSecret,
+      redirect_uris: [callbackUrl2],
```

Comment:
personally I just find the variable name yucky
```suggestion
      redirect_uris: [customCallbackUrl || callbackUrl],
```